### PR TITLE
Enables gogoproto custom type Key to be used with auto-generated marshal.

### PR DESCRIPTION
**Pull request by BramGruneir**
This also fixes a bug in (k *Key) Unmarshal(bytes []byte).

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM.

File: proto/data.go [hunk 1]
```diff
@@ -184,9 +184,17 @@ func (k EncodedKey) Marshal() ([]byte, error) {
...
-// Unmarshal implements the gogoproto Unmarshaler interface.
+// Unmarshal implements the gogoproto Unmarshaler interface.  It should never
+// set k to be nil as a nil value should never have unmarshall called for it.
+// Futhermore, appending a 0 length byte slice to a 0 length byte slice will
+// produce a nil byte slice, which can cause problems when unmarshalling the
+// engine.MinKey of "".
 func (k *Key) Unmarshal(bytes []byte) error {
-	*k = Key(append([]byte(nil), bytes...))
+	if len(bytes) == 0 {
+		*k = []byte{}
```
Comment (petermattis) [inline proto/data.go]: I'm surprised you don't have to cast to `Key` here.

**Comment by petermattis** [comment]
LGTM

File: proto/data.go [hunk 2]
```diff
@@ -184,9 +184,17 @@ func (k EncodedKey) Marshal() ([]byte, error) {
 	return []byte(k), nil
 }
 
-// Unmarshal implements the gogoproto Unmarshaler interface.
+// Unmarshal implements the gogoproto Unmarshaler interface.  It should never
+// set k to be nil as a nil value should never have unmarshall called for it.
```
Comment (spencerkimball) [inline proto/data.go]: s/unmarshall/Unmarshal/

File: proto/config.proto [hunk 3]
```diff
@@ -103,6 +103,9 @@ message RangeTree {
 
 // RangeTreeNode holds the configuration for each node of the Red-Black Tree that references all ranges.
 message RangeTreeNode {
+  option (gogoproto.marshaler) = true;
```
Comment (spencerkimball) [inline proto/config.proto]: Add a comment here describing why we need these generated.

**Comment by spencerkimball** [comment]
LGTM

**Pull request merged**